mdx/guides/develop-a-margin-trading-smart-contract-with-0x-api.mdx

```diff
@@ -0,0 +1,573 @@
+
+```solidity
+function open(ZeroExQuote memory quote)
+  public
+  payable
+  onlyOwner
+  onlyWhenClosed
+  returns (uint256 positionBalance, uint256 wethBalance, uint256 borrowBalance)
+{
+    ...
+  // 1. increase position by msg.value - protocolFeeAmount
+  positionBalance = msg.value.safeSub(quote.protocolFeeAmount);
+}
+```
+
+#### Deposit and borrow DAI
+
+With `positionBalance`, we can then deposit and mint cETH with Compound Finance.
+
+```solidity
+...
+// 2. increase position by msg.value - protocolFeeAmount
+positionBalance = msg.value.safeSub(quote.protocolFeeAmount);
+// 3. mint collateral in compound
+CETH.mint.value(positionBalance)()
+...
+```
+
+Now that we have `accountLiquidity` (borrowing power in Compound Finance), let's borrow DAI from Compound Finance. We will be borrowing the exact amount of DAI we intend to sell with 0x liquidity. Off-chain, `sellAmount` is calculated based on our desired leverage.
+
+```
+...
+// 4. borrow token
+assert(CDAI.borrow(quote.sellAmount) == 0);
+...
+```
+
+Compound finance returns a non-zero value if an error occurs during the function call, with an `assert` we can proceed with the function only when the function doesn't throw any errors.
+
+#### Swaping for ETH with 0x liquidity 
+
+To swap with 0x, the 0x exchange contracts will need to have access to move DAI from the `MarginPositionContract`. With the provided functions in the starter code, let’s approve `MAX_AMOUNT` of DAI: 
+
+```solidity
+...
+// 4. approve 0x exchange to move DAI
+_approve(address(DAI), _getZeroExApprovalAddress());
+...
+```
+
+Since the quote is passed off-chain, it is worth verifying that the assets you intend to sell and buy.
+
+```solidity
+...
+// 5. verify quote is valid
+require(quote.sellToken == address(WETH), "not buying WETH");
+require(quote.buyToken == address(DAI), "not buying WETH");
+...
+```
+
+Now that we have verified and set the approvals to perform a swap with 0x protocol, let’s call `EXCHANGE` with `callDataHex` provided in the quote.
+
+```solidity
+...
+// 6. execute swap
+(bool success, bytes memory data) = address(EXCHANGE).call.value(quote.protocolFeeAmount)(quote.calldataHex);
+require(success, "Swap not filled.");
+// 7. decode fill results
+LibFillResults.FillResults memory fillResults = abi.decode(data (LibFillResults.FillResults));
+...
+```
+
+#### Updating the contract state
+
+With that, we have just swapped DAI for WETH, effectively opening our margin position. Before we finish this function, let’s update the contract state and return some results to the user. From the 0x contract call, we have a receipt that contains the actual amount of buy and sell token swapped.
+
+```solidity
+...
+// 8. position size increase by bought amount of WETH (makerAssetFilledAmount)
+positionBalance += fillResults.makerAssetFilledAmount;
```

Comment:
We should never use arithmetic operators. Let's use `positionBalance = positionBalance.safeAdd(fillResults.makerAssetFilledAmount)` here.